rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,635 @@
+       entry ``000X`` as this would "hide" ``0001``.
+    b) would not cause an entry "contained" within an entry of higher
+       generality to be hidden by the insertion of a new entry. For example, if
+       the entry ``XXXX`` had been formed by merging the entries ``0011`` and
+       ``1100`` then merging of the entries ``1101`` and ``1110`` would not be
+       allowed as it would cause the entry ``11XX`` to be inserted above
+       ``XXXX`` in the table and would hide ``1100``.
+
+Following these rules ensures that the minimised table will be functionally
+equivalent to the original table provided that the original table was invariant
+under reordering OR was provided in increasing order of generality.
+"""
+from collections import defaultdict, namedtuple
+from rig.routing_table import MinimisationFailedError, RoutingTableEntry
+from six import itervalues
+
+
+def minimise(routing_table, target_length=1024):
+    """Reduce the size of a routing table by merging together entries where
+    possible.
+
+    .. warning::
+
+        The input routing table *must* also include entries which could be
+        removed and replaced by default routing.
+
+    .. warning::
+
+        It is assumed that the input routing table is not in any particular
+        order and may be reordered into ascending order of generality without
+        affecting routing correctness.  It is also assumed that if this table
+        is unordered it is at least orthogonal (i.e., there are no two entries
+        which would match the same key) -- this assumption is not checked!
+
+        .. note::
+
+            If *all* the keys in the table are derived from a single instance
+            of :py:class:`~rig.bitfield.BitField` then the table is guaranteed
+            to be orthogonal.
+
+        .. note::
+
+            Use :py:meth:`~rig.routing_table.expand_entries` to generate an
+            orthogonal table and receive warnings if the input table were not
+            orthogonal.
+
+    Parameters
+    ----------
+    routing_table : [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+        Routing entries to be merged.
+
+    Other Parameters
+    ----------------
+    target_length : int or None
+        Target length of the routing table; the minimisation procedure will
+        halt once either this target is reached or no further minimisation is
+        possible. If None then the table will be made as small as possible.
+
+    Raises
+    ------
+    MinimisationFailedError
+        If the smallest table that can be produced is larger than
+        `target_length`.
+
+    Returns
+    -------
+    [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+        Reduced routing table entries.
+    """
+    # We maintain a dictionary of which keys and masks in the routing table are
+    # combinations of other keys and masks; this allows us to be less
+    # aggressive in determining whether inserting a new entry into the table
```

Comment:
"We maintain a dictionary of which keys and masks in the routing table are combinations of other (now removed) keys and masks; this allows us to consider only the keys and masks the user actually cares about when determining if inserting a new entry will break the correctness of the table."